gcovr/gcov_parser.py

```diff
@@ -0,0 +1,994 @@
+_RE_UNCONDITIONAL_LINE = _line_pattern(
+    r"unconditional (INT) (?:taken (VALUE)|never executed)"
+)
+_RE_BLOCK_LINE = _line_pattern(r"(?: )?(VALUE|[$]{5}|[%]{5}): (INT)-block (INT)")
+
+
+class _ExtraInfo(enum.Flag):
+    """Additional info about lines, such as noncode or exception-only status."""
+
+    NONE = 0
+    NONCODE = enum.auto()
+    EXCEPTION_ONLY = enum.auto()
+    PARTIAL = enum.auto()
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+class _SourceLine(NamedTuple):
+    """A gcov line with source code: ``HITS: LINENO:CODE``"""
+
+    hits: int
+    lineno: int
+    source_code: str
+    extra_info: _ExtraInfo
+
+
+class _MetadataLine(NamedTuple):
+    """A gcov line with metadata: ``-: 0:KEY:VALUE``"""
+
+    key: str
+    value: str
+
+
+class _BlockLine(NamedTuple):
+    """A gcov line with block data: ``HITS: LINENO-block BLOCKNO``"""
+
+    hits: int
+    lineno: int
+    blockno: int
+    extra_info: _ExtraInfo
+
+
+class _SpecializationMarkerLine(NamedTuple):
+    """A gcov line that delimits template specialization sections (no fields)"""
+
+
+class _SpecializationNameLine(NamedTuple):
+    """A gcov line with the name of a specialization section: ``NAME:``"""
+
+    name: str
+
+
+class _CallLine(NamedTuple):
+    """A gcov line with call data: ``call CALLNO returned RETURNED``"""
+
+    callno: int
+    returned: int
+
+
+class _BranchLine(NamedTuple):
+    """A gcov line with branch data: ``branch BRANCHNO taken HITS (ANNOTATION)``"""
+
+    branchno: int
+    hits: int
+    annotation: Optional[str]
+
+
+class _UnconditionalLine(NamedTuple):
+    """A gcov line with unconditional branch data: ``unconditional BRANCHNO taken HITS``"""
+
+    branchno: int
+    hits: int
+
+
+class _FunctionLine(NamedTuple):
+    """
+    A gcov line with function coverage data for the next line.
+
+    ``function NAME called CALLS returned RETURNED blocks executed BLOCKS``
+    """
+
+    name: str
+    calls: int
+    returned: int
+    blocks_covered: int
+
+
+# NamedTuples can't inherit from a common base,
+# so we use a Union type as the parse_line() return type.
+#
+# Why NamedTuples? Better type safety than tuples, but very low memory overhead.
+_Line = Union[
+    _SourceLine,
+    _MetadataLine,
+    _BlockLine,
+    _SpecializationMarkerLine,
+    _SpecializationNameLine,
+    _CallLine,
+    _BranchLine,
+    _UnconditionalLine,
+    _FunctionLine,
+]
+
+
+class UnknownLineType(Exception):
+    """Used by `parse_line()` to signal that no known line type matched."""
+
+
+def parse_metadata(lines: List[str]) -> Dict[str, str]:
+    r"""
+    Collect the header/metadata lines from a gcov file.
+
+    Example:
+    >>> parse_metadata('''
+    ...   -: 0:Foo:bar
+    ...   -: 0:Key:123
+    ... '''.splitlines())
+    {'foo': 'bar', 'key': '123'}
+    """
+    collected = dict()
+    for line in lines:
+        # IO.readlines() leaves trailing newlines...
+        line = line.rstrip("\n")
```

Comment:
Could the rstrip in each for loop be moved to the place where the file is read?

Reply:
Great point! Now, the gcov module does `INPUT.read().splitlines()` instead of `INPUT.readlines()` which takes care of the problem at the source.